matrix_sdk/src/room/common.rs

```diff
@@ -146,6 +147,17 @@ impl Common {
         self.client.send(request, None).await
     }
 
+    /// Sends a request to `/_matrix/client/r0/rooms/{roomId}/event/{eventId}`
+    /// and returns a `get_room_event::Response` that contains a event
+    /// (`AnyRoomEvent`).
+    pub async fn event(
+        &self,
+        request: impl Into<get_room_event::Request<'_>>,
+    ) -> Result<get_room_event::Response> {
+        let request = request.into();
+        self.client.send(request, None).await
```

Comment:
This should check if the event is encrypted and decrypt it if it is. We'll need to introduce a new custom event type for this to work correctly and not lose info about the encryption state of the event.

It should be similar to the `SyncRoomEvent` we have in the [deserialized_responses](https://github.com/matrix-org/matrix-rust-sdk/blob/4432be48e060227ffc6df5079f494aaf2c49a063/matrix_sdk_common/src/deserialized_responses.rs#L90) module.

A method will need to be added to the `Client` so we can decrypt such an event:

```rust
async fn decrypt_room_event(&self, event: &AnyMessageEvent) {
    if let Some(machine) = self.base_client.olm_machine().await {
        if let AnyMessageEvent::RoomEncrypted(e) = event {
            let room_id = event.room_id();
            // Turn the AnyMessageEvent into a AnySyncMessageEvent
            let event = event.clone().into();

            if let AnySyncMessageEvent::RoomEncrypted(e) = event {
                let decrypted = machine.decrypt_room_event(&e, &room_id).await;
            };
        }
    }
}
```

Reply:
Oh I didn't know that `Client::send()` can be used for that too. I have few questions:
1) What should `decrypt_room_event()` return? Your code snippet doesn't seem to make sense for me without a return value. What about `Result<SyncRoomEvent, MegolmError>` ?
2) Why the need for a new custom event type? How it should look like? Why we cannot use the already existing `SyncRoomEvent` type? 
3) Can you elaborate what do you mean with "not lose info about the encryption state of the event."?
4) I guess the return value of `event()` should be the not existing yet custom event type?

Sorry about the many questions, I'm not that familiar with the matrix-rust-sdk internals yet. I didn't thought about encryption in this PR, and thought it would be a beginner-friendly low hanging fruit to implement this :)

Reply:
>     1. What should `decrypt_room_event()` return? Your code snippet doesn't seem to make sense for me without a return value. What about `Result<SyncRoomEvent, MegolmError>` ?

I think returning just the custom event type makes sense, we can ignore `MegolmError`s since that's only notifying us that decryption failed, in that case we'll return the undecrypted event.

>     2. Why the need for a new custom event type? How it should look like? Why we cannot use the already existing `SyncRoomEvent` type?

We need a custom event type because we can't attach the encryption info to the existing event types that we use from Ruma. Granted we could return a tuple here instead, but since we're already using and going to such a custom wrapper type it makes sense to stay consistent.

It should look like:

```rust
pub struct RoomEvent {
    /// The actual event.
    pub event: Raw<ruma::events::AnyRoomEvent>,
    /// The encryption info about the event. Will be `None` if the event was not
    /// encrypted.
    pub encryption_info: Option<EncryptionInfo>,
}
```

We can't reuse `SyncRoomEvent` because sync responses return a different enum, `AnySyncRoomEvent` vs `AnyRoomEvent`.

>     3. Can you elaborate what do you mean with "not lose info about the encryption state of the event."?

I mean by that that clients want to know if an event was encrypted or not, thus keeping the `EncryptionInfo` around after we decrypt an event.

>     4. I guess the return value of `event()` should be the not existing yet custom event type?

Correct.

> Sorry about the many questions, I'm not that familiar with the matrix-rust-sdk internals yet. I didn't thought about encryption in this PR, and thought it would be a beginner-friendly low hanging fruit to implement this :)

Sorry for catching you off guard, though it's still quite an easy PR. The most complex thing is adding the decryption method that was posted above, after that it's just connecting all this.

Reply:
Thanks for the help, I updated the PR. One thing I noticed: Wouldn't it be better to take a `AnyRoomEvent` for `decrypt_room_event()`? We want to limit it to room events either way, and it allows us to return the unencrypted event more easily.

Reply:
Sure, the exact event that `decrypt_room_event()` takes isn't that important, it might make most sense to let it take `MessageEvent<EncryptedEventContent>` which is the fully deconstructed version of it. But I'm not sure that you can convert it to the sync variant once you deconstruct it.